### DISCUSSION TITLE
Could not get correct stress-strain respnse using powerlaw constituve in damask3.0.0

### DISCUSSION THREAD

**Discussion by ruanqc** [damask-multiphysics/DAMASK#46] [Q&A] [answered]
Hello DAMASK team,
I am learning use DAMASK v3.0.0 to reproduce work ‘A spectral method solution to crystal elasto-viscoplasticity at finite strain’ by Prof. Eisenlohr et a.., published in IJP 2013. ( I have run it before in damask2.0.3 and got the right results in single crystal cases(compression in z-direction with 40% strain). While in the current damask3.0.0 version, I can not get the correct equivalent stresss-strain curve and it seams that no plastic response, although the displacement shows strain reaches 40%. 
I am not sure whether the problem comes from the grid, or material orientation quaternion parameters? I'd very appriciated if anyone could give some advices. 

1. I generated the grid through
“import damask
import numpy as np
size = np.ones(3)  
cells = [8,8,8] 
N_grains = 1 
seeds = damask.seeds.from_random(size,N_grains,cells)
grid = damask.GeomGrid.from_Voronoi_tessellation(cells,size,seeds) 
grid.save(f'geom_N{N_grains}_{cells[0]}x{cells[1]}x{cells[2]}') 
grid
“

2. I change the orientation as -O[1.0,0,0,0] O: [1.0, 0.0, 0.0, 0.0], which corresponds to a respect Euler angles phi1=0, phi=0, and phi1=0. 
“
---
homogenization:
  SX:
    N_constituents: 1
    mechanical: {type: pass}
phase:
  Aluminum:
    lattice: cF
    mechanical:
      output: [F, P, F_e, F_p, L_p, O]
      elastic: {type: Hooke, C_11: 106.75e+9, C_12: 60.41e+9, C_44: 28.34e+9}
      plastic:
        output: [xi_sl]
        type: phenopowerlaw
        N_sl: [12]
        a_sl: [2.25]
        atol_xi: 1.0
        dot_gamma_0_sl: [0.001]
        h_0_sl-sl: [75.e+6]
        h_sl-sl: [1, 1, 1.4, 1.4, 1.4, 1.4, 1.4]
        n_sl: [20]
        xi_0_sl: [31.e+6]
        xi_inf_sl: [63.e+6]
material:
  - homogenization: SX
    constituents:
      - phase: Aluminum
        v: 1.0
        O: [1.0, 0.0, 0.0, 0.0]
"

3. I plot the equivalent stress-strain through
“
import damask
import numpy as np
import matplotlib.pyplot as plt
res = damask.Result('outdata_phi0.hdf5')
res.add_stress_Cauchy()
res.add_strain()
res.add_equivalent_Mises('sigma')
res.add_equivalent_Mises('epsilon_V^0.0(F)')
sigma = [-np.average(s)/1.e6 for s in res.get('sigma').values()]  
epsilon=[-np.average(e)*100 for e in  res.get('epsilon_V^0.0(F)').values()] 
plt.plot(epsilon,sigma)
“
and the results shown as ,
![image](https://github.com/user-attachments/assets/93b15d11-ab84-4230-80e3-f33a06bf1472)
![image](https://github.com/user-attachments/assets/d2de8008-b457-45a9-a029-10a2f684e782)

I attached the input file in the .zip file in case you may use them. 
[Input_file_powerLaw.zip](https://github.com/user-attachments/files/17995924/Input_file_powerLaw.zip)

Thank you and best regards,  
ruanqc

**Comment by MarDiehl** [comment, marked as answer]
it's just an issue of getting the data. You use the full stress and strain tensor and average over all points and all components. The datasets with the equivalent quantities are suffixed by `_vM`, so 

```Python
import damask
import numpy as np
import matplotlib.pyplot as plt
res = damask.Result('geom_N1_8x8x8_compressionZ_n1_material.hdf5')
res.add_stress_Cauchy()
res.add_strain()
res.add_equivalent_Mises('sigma')
res.add_equivalent_Mises('epsilon_V^0.0(F)')
sigma = [np.average(s)/1.e6 for s in res.get('sigma_vM').values()]
epsilon=[np.average(e)*100 for e in res.get('epsilon_V^0.0(F)_vM').values()]
plt.plot(epsilon,sigma)
plt.show()
```
works

**Reply by ruanqc** [reply]
I see! Sorry Prof. Diehl for my careless mistake and thank you for your helpful answer.

Another question is that whether DAMASK have some relative instructions on how to implement specific material model into damask. whether by adding constitutive model OR by modifing one of those already esisted in the source code.  I have try to modied the source code, but it seams not work correctly.

For a testing try, 
     1. I firstly copied "phase_mechanical_plastic_isotropic.f90" and rename a new model called "phase_mechanical_plastic_new.f90", then modified it by replacing 'isotropic' with 'new',
then i referred isotropic one to added subroutine interfaces  in new phase_mechanical_plastic_new.f90 into the following source codes:
damask/src/phase_mechanical_plastic.f90
damask/src/phase_mechanical.f90
damask/src/phase.f90
    2. after that I recompiled to install the DAMASK_grid solver, and read compiled successfully.
"
[100%] Linking Fortran executable DAMASK_grid
[100%] Built target DAMASK_grid
Install the project...
-- Install configuration: "RELEASE"
-- Installing: /usr/local/bin/DAMASK_grid
"
[new_isotropic.zip](https://github.com/user-attachments/files/18024345/new_isotropic.zip)

    3. However, when I used this as in n1_new_powerLaw.yaml , 
      plastic:
        type: new

it seams that damask could not reconised this new adding model and got error called "201, unkown plasticity specified". 

So I suppose there should be some other interfaces or source files that required to modifed?  I'd very appreciated if you could give some advices on these, or could I ask for help though email (damask@mpie.de)? At the end of day, I really do want to implement my constitutive model that much similar to that existed nonlocal one (Originally probposed in Dr. Christoph Kords's PhD Dissertation, and implemented by Prof. Diehl you and Prof. Eisenlohr, et al. ) 

Thank you and best regards, ruanqc

**Reply by MarDiehl** [reply]
Copying an existing model is a good idea. For development, it might even easier to modify a model.

In case that you create a copy, you need to add the functions of the new model to several `switch/case` statements in `phase_mechanical_plastic.f90`, maybe also some to `phase_mechanical_elastic.f90` or `phase_mechanical.f90`. Just search for something like `dislotwin` in all files. In the plasticity model, there is also a statement like `myPlasticity = plastic_active('isotropic')` where you need to have a different label.

**Reply by ruanqc** [reply]
> In case that you create a copy, you need to add the functions of the new model to several `switch/case` statements in `phase_mechanical_plastic.f90`, maybe also some to `phase_mechanical_elastic.f90` or `phase_mechanical.f90`. Just search for something like `dislotwin` in all files. In the plasticity model, there is also a statement like `myPlasticity = plastic_active('isotropic')` where you need to have a different label.

I basically changed it according to this idea before, may be some places have not been searched and changed. 
Thank you so much professor.
